# Рекомендательные системы (РС)

Самый простой пример РС, с которым сталкивались практически все, — это системы рекомендации фильмов.

Например, такая есть у сайта kinopoisk.ru: пользователи сервиса выставляют оценки просмотренным фильмам и отмечают те, которые им не понравились. После получения некоторой доли оценок сайт предлагает киноленты, которые, скорее всего, понравятся конкретному пользователю. Похожие РС системы существуют практически везде, от музыки до интернет-маркетов.

Методы построения рекомендательных систем можно представить следующим образом:

<img src='Images/math_ml_22.png'>

# Рекомендательные системы

### Персонализированные

Строят рекомендации индивидуально для пользователя.

* На основе контента

Рекомендуют объекты, похожие на те, что пользователь уже оценил (анализируют свойства объектов).

* Коллаборативная фильтрация

Использует поведение других пользователей.

* На основе памяти

На основе пользователей — ищут похожих пользователей и рекомендуют их предпочтения.

На основе элементов — рекомендуют элементы, похожие на те, что уже понравились пользователю.

* На основе моделей

Строят модель предсказания оценок на основе данных (например, регрессия).

Матричная факторизация - Разлагают матрицу пользователь-объект (например, SVD) для выявления скрытых факторов.

* Гибридные - Комбинируют несколько методов (например, контент + коллаборативная).

### Неперсонализированные

Дают одинаковые рекомендации всем.

* На основе популярности - Рекомендуют самые популярные товары/фильмы/книги и т.д.

# Данные для рекомендательной системы

Существует два способа сбора данных для рекомендательных систем — явный (англ. explicit feedback, explicit ratings) и неявный (англ. implicit feedback, implicit ratings).

# Явный сбор данных (explicit feedback)

Получают ту информацию, которую передают сами пользователи, например это оценки фильмов на Netflix или товаров на Amazon. Явные отзывы учитывают мнение пользователя о том, насколько ему понравился или не понравился продукт. Данные явных отзывов легко поддаются количественной оценке.

### Однако использование таких данных сопряжено с рядом проблем:

* После сбора явных отзывов вы с большой вероятностью получите набор оценок с __ярко выраженной полярностью__ (очень часто люди ставят оценку только в случае, если что-либо вызвало у них очень сильные эмоции, преимущественно негативные).

* Явные отзывы __не учитывают контекст__, в котором был оценён тот или иной продукт.

* Также есть __сложности с выстраиванием системы оценок__: необходимо придумать такую шкалу, чтобы пользователям было легко выражать своё мнение и это действие не вызывало у них раздражение.

# Неявный сбор данных (implicit feedback)

Неявная обратная связь не отражает непосредственный интерес пользователя, но действует как косвенный показатель этого интереса.

Примерами неявных данных являются история просмотров, клики по ссылкам, подсчёт количества проигрываний песни, процент прокрутки веб-страницы или даже движение курсора по странице.

<img src='Images/math_ml_23.png'>

### Также стоит упомянуть о некоторых особенностях неявных данных:

* Отсутствие прямого измерения __негативных предпочтений__.

В отличие от явных отзывов, когда пользователь ставит плохую оценку, у нас нет прямого способа измерить негативное отношение пользователя к продукту.

* Возможность делать верные выводы по численному выражению обратной связи.

Если выразить неявную обратную связь в численном эквиваленте, мы сможем однозначно определять, насколько какой-то товар или контент привлекают пользователя.

* Большой объём зашумлённых данных.

Прежде чем использовать данные неявной обратной связи для построения рекомендательной системы, придётся потратить много времени на их предобработку.

# Метрики в рекомендательных системах

Все метрики для оценки качества РС можно разделить на две группы:

* офлайн-метрики (оценивают качество алгоритма);
* онлайн-метрики (оценивают производительность и бизнес-показатели).

# Офлайн-метрики

Существует несколько категорий офлайн-метрик:

- __Prediction Accuracy__ — оценка точности предсказываемого рейтинга.
- __Decision Support__ — оценка релевантности рекомендаций.
- __Rank Accuracy__ — оценка качества рекомендаций с учётом ранжирования.

### Prediction Accuracy

Метрики __Prediction Accuracy__ сравнивают прогнозируемые значения рейтинга с реальными оценками пользователей. Для того чтобы количественно оценить схожесть фактических и предсказанных данных, обычно используются уже знакомые нам метрики MAE, MSE и RMSE.

Примечание: Чем ниже MAE и RMSE, тем лучше, ведь они дают нам представление о том, насколько точны наши прогнозные оценки и рекомендации.

Для того чтобы вычислить MAE и RMSE при построении рекомендательной системы, можно воспользоваться уже известными вам готовыми функциями из модуля sklearn:

```python
 sklearn.metrics.mean_absolute_error(y_true, y_pred)
 sklearn.metrics.mean_squared_error(y_true, y_pred)
```

### Decision Support

Следующие метрики — это уже знакомые нам по задаче классификации RECALL и PRECISION.

К сожалению, в известных нам библиотеках для машинного обучения эти функции не реализованы. Для оценки качества рекомендуется использовать модуль cute_ranking.

### Rank Accuracy

Рекомендательные системы должны быть способны помещать __наиболее релевантные элементы__(те, которые клиент оценил или в итоге купил) на высокие позиции в списке рекомендаций. Первый рекомендуемый элемент всегда самый важный, следующий — второй по важности и т.д.

Сейчас рассмотрим метрики, которые как раз учитывают положение продукта в рейтинге:

* MRR (Mean Reciprocal Rank);
* MAP (Mean Average Precision);
* NDCG (Normalized Discounted Cumulative Gain).

### MRR (средний реципрокный ранг)

Это самая простая метрика из трёх перечисленных. Она определяет, где среди рекомендованных находится первый релевантный элемент.

<img src='Images/math_ml_24.png'>

### Преимущества

* Метод прост в вычислениях и легко интерпретируется.
* Метод уделяет большее внимание первому релевантному элементу списка, что в целом отражает логику рекомендательных систем.

### Недостатки

* Метрика фокусируется на одном элементе из списка и не оценивает остальные рекомендуемые элементы.
* Список с одним релевантным элементом имеет такой же вес в вычислении итогового показателя, как и список с большим количеством релевантных элементов. Это не всегда хорошо.
* Метрика плохо подходит для случаев, когда важно получить именно ряд рекомендаций, а не одну рекомендацию.

Для вычисления MRR в Python рекомендуем использовать функцию mean_reciprocal_rank() из библиотеки cute_ranking.

### MAP (средняя точность)

Это средняя точность по всем релевантным объектам в списке рекомендаций.

Он показывает: насколько хорошо система расставила релевантные объекты в списке — чем выше релевантные объекты в списке, тем лучше MAP.

<img src='Images/math_ml_25.png'>

### Преимущества

* Метрика естественным образом обрабатывает ранжирование списков рекомендованных элементов.
* Метрика способна придавать вес ошибкам пропорционально их месту в списке: больший вес — ошибкам в верхней части, меньший вес — ошибкам ниже по списку. Это соответствует необходимости показать как можно больше релевантных элементов в верхней части списка рекомендаций.

### Недостатки

* Метрика отлично подходит для бинарных (релевантных/нерелевантных) оценок, однако не подходит для рейтинговых числовых оценок.

### NDCG (нормализованный дисконтированный кумулятивный выигрыш)

Подход к вычислению этой метрики схож с подходом к вычислению MAP: в обоих случаях высоко ценится выдача релевантных продуктов на первых позициях. Однако метрика NDCG позволяет использовать информацию о том, что разные продукты имеют разную степень релевантности, то есть учитывает их рейтинг!

Представим, что поисковая система выдаёт пять статей с именами D1, D2, D3, D4, D5, которые выводятся в таком же порядке. Определим шкалу релевантности (0–3), где:

* 0 — не релевантно;
* 1–2 — в некоторой степени релевантно;
* 3 — полностью релевантно.

Предположим, статьи имеют следующие оценки релевантности:

* D1 — 3;
* D2 — 2;
* D3 — 0;
* D4 — 0;
* D5 — 1.

В модуле sklearn есть готовая функция:

```python
    from sklearn.metrics import ndcg_score, dcg_score
    import numpy as np

    true = np.asarray([[3, 2, 1, 0, 0]])
    relevance = np.asarray([[3, 2, 0, 0, 1]])

    print(ndcg_score(true, relevance))

    # 0.980840401274087
```

# Онлайн-метрики

Выделяют пять групп бизнес-показателей, на которые рекомендательные системы оказывают самое существенное влияние:

1. CTR (Click-Through Rate)

Доля кликов по рекомендованным товарам (показывает привлекательность рекомендаций для пользователя).
Пример: сколько раз пользователи кликнули на товар из рекомендаций относительно числа показов.

2. Применение и конверсия

Как часто рекомендации приводят к целевому действию (например, к покупке, добавлению в корзину, регистрации).
Пример: сколько рекомендованных товаров куплено после клика.

3. Продажи и выручка

Вклад рекомендаций в общий объём продаж и доход компании.
Пример: сколько процентов от общей выручки принесли рекомендованные товары.

4. Охват различных категорий товара

Насколько рекомендации помогают пользователям открывать новые категории или ассортимент.
Пример: рост продаж в ранее не популярных категориях.

5. Поведение пользователей

Как рекомендации влияют на взаимодействие с платформой (удержание, вовлечённость).
Пример: увеличение времени на сайте, числа возвращений пользователей.

# Проблема холодного старта и popularity-based model

Рекомендательные системы строятся на данных о товарах и о пользователях. Но что делать, если какой-то информации не хватает или если данных вообще нет?

В таком случае говорят о проблеме холодного старта. Требуется потратить немного времени или сделать несколько шагов, чтобы «разогреть» сервис рекомендаций, прежде чем он начнёт выдавать релевантные результаты с достаточной эффективностью.

Можно выделить два вида проблем холодного старта:

* проблема холодного старта пользователя;
* проблема холодного старта продукта.

### Холодный старт пользователя

Когда система сталкивается с новыми посетителями веб-сайта, не имеющими истории просмотров или известных предпочтений, создание персонализированного опыта для них становится сложной задачей, поскольку данные, обычно используемые для создания рекомендаций, отсутствуют.

Зачастую основной метод при работе с новыми пользователями — попросить их в явном виде сообщить свои предпочтения, чтобы создать первоначальный профиль пользователя. Вы могли видеть формат такого взаимодействия на Netflix, когда про интересы и любимые сериалы спрашивают при регистрации!

### Холодный старт продукта

Когда новый товар добавляется в интернет-магазин или когда свежий контент загружается на медиаплатформу, первое время о нём никто не знает. С нулевым количеством взаимодействий или оценок он практически невидим для рекомендательной системы независимо от того, насколько релевантным он будет для пользователей.

Среди тех, кто больше всего страдает от этого явления, — сайты объявлений и новостные платформы. Здесь свежие продукты, как правило, самые «горячие», но их ценность быстро снижается: вчерашняя новость сегодня уже устарела, а велосипед, выставленный на продажу на прошлой неделе, уже продан.

### Popularity-based model

Cейчас рассмотрим один из вариантов построения рекомендаций в случае холодного старта пользователя.

Если к нам приходит клиент, про которого мы ничего не знаем, мы можем создать для него рекомендации с использованием popularity-based-модели. Это тип РС, которая формирует рекомендации на основе популярности продуктов. Например, если большинство людей часто покупает определённый продукт, то система знает, что он наиболее востребован среди аудитории, поэтому каждому новому пользователю, который только что зарегистрировался, система будет рекомендовать __этот продукт__.

### Преимущества

* Не страдает от проблем холодного старта.
* Нет необходимости в исторических данных для пользователя.

### Недостатки

* Не персонализирована (будет рекомендовать одинаковые продукты всем пользователям).

Примеры использования popularity-based-моделей в известных сервисах:

* Google Новости — новости, отфильтрованные по трендам и самым популярным новостям;
* YouTube — самые популярные видео.



# Модель на основе контента (Content based)

Это модель персонализированная (personalized)

Такие рекомендательные системы базируются на детальном анализе данных пользователей, изучении их действий, покупок, взаимодействия с другими пользователями. Таким образом, каждый человек получает индивидуальную рекомендацию конкретно под его особенности и предпочтения.

Наиболее популярными типами персонализированных рекомендательных систем являются системы, основанные на __контенте (content-based)__ и __коллаборативной фильтрации (collaborative filtering)__.

Рекомендательные системы на основе контента используют метаданные (дополнительные данные о данных) о товарах или пользователях для создания конкретных рекомендаций. Например, если пользователь уже читал книгу одного автора или покупал товар определённой марки, то предполагается, что он отдаёт предпочтение этому автору или этой марке, и существует вероятность, что он купит аналогичный товар в будущем.

Предположим, пользователю Ане нравятся научно-фантастические книги, и её любимый писатель — Уолтер Джон Уильямс. Если она прочитала книгу "Aristoi", то рекомендуемой ей книгой будет "Angel Station" — ещё одна научно-фантастическая книга, написанная Уолтером Джоном Уильямсом.

<img src='Images/math_ml_26.png'>

Примечание: Сходство или близость элементов измеряется на основе сходства содержания этих элементов. Говоря «содержание», имеется в виду такие сущности, как категория, тег, жанр и т. д., то есть метаданные.

Рассмотрим два наиболее популярных метода измерения схожести:

* индекс Жаккара,
* косинусная близость.

### Индекс Жаккара

__Индекс Жаккара__ измеряет сходство между двумя наборами A и B как мощность множества пересечения, делённую на мощность множества объединения каких-то характеристик объекта. Его удобно применять для категориальных признаков.

Например, с помощью индекса Жаккара мы можем оценить, насколько похожи фильмы, основываясь на наборах ключевых слов (тегов) для них:

* фильм А : {фантастика, школа, романтика};
* фильм B : {приключения, фантастика, школа};
* фильм C : {ужасы, триллер, драма}.

Мы можем предположить, что фильм A больше похож на фильм B, чем на фильм C, так как фильмы A и B имеют два общих тега (фантастика, школа), в то время как фильмы A и C не имеют ни одного общего тега.

Если бы мы рассматривали вычисление этого индекса для фильмов А и В, то получили бы $\frac{2}{4}$, так как в пересечении два тега, а в объединении — четыре. Для фильмов А и C мы получили бы 0, так как пересечение множеств их тегов является пустым. Таким образом, показатель близости для фильмов А и В получился бы больше, и это подтвердило бы наши предположения об их большей схожести.

### Косинусная близость

Подход с использованием индекса Жаккара помог нам создать интуитивное представление о том, что означает сходство набора категориальных значений. Подход с косинусным сходством немного сложнее и применяется для оценки близости массивов с числами. Он требует, чтобы мы представляли объекты в виде __вектора__.

Например, мы можем представить те же самые фильмы как набор из трёх вещественных чисел:

* фильм A = (1.1, 2.3, 5.1);
* фильм B = (1.3, 2.1, 4.9);
* фильм C = (5.1, 6.2, 1.1).

В этом случае, по сути, мы ищем угол между векторами. Если два вектора совпадут (т.е. будут максимально близкими), то угол между ними будет равен нулю, а значит, косинус будет равен 1. Если векторы будут направлены в противоположные стороны, косинус будет равен -1. Таким образом, мы можем получить для любых двух векторов значение в пределах от -1 до 1 включительно, по которому можно определить, насколько векторы близки друг к другу.

Вычислим косинусную близость между векторами А и С. Результат округлим до трёх знаков после точки-разделителя.

In [2]:
import numpy as np

# Векторы
a = np.array([1.1, 2.3, 5.1])
c = np.array([5.1, 6.2, 1.1])

# Скалярное произведение
dot_product = np.dot(a, c)

# Нормы (длины)
norm_a = np.linalg.norm(a)
norm_c = np.linalg.norm(c)

# Косинусная близость
cosine_similarity = dot_product / (norm_a * norm_c)
print(f"Косинусная близость: {cosine_similarity:.3f}")

Косинусная близость: 0.551


Как нам известно, чем выше значение косинусной близости, тем больше сходство между векторами. Таким образом, наши изначальные предположения совпали с реальностью: векторы А и В действительно схожи между собой сильнее (косинусная близость 0.999), чем векторы А и С (0.551). Значит, если человек поставил высокую оценку фильму А, то мы должны порекомендовать ему фильм В.

Итак, чтобы построить рекомендательную систему на основе контента, необходимо:

1. Для каждого продукта создать характеризующие его признаки.
2. Найти показатель близости между всеми продуктами.
3. Порекомендовать пользователю продукты, которые показывают наибольшую близость с теми продуктами, которые он высоко оценил.

### Преимущества

* __Для создания рекомендаций не требуются данные от других пользователей__. Как только пользователь выполнил поиск, просмотрел несколько продуктов и/или совершил несколько покупок, система фильтрации на основе контента может начать создавать соответствующие рекомендации. Это делает её идеальной для компаний и сервисов, у которых нет огромного количества пользователей для формирования выборки.
* __Рекомендации получаются очень релевантными для пользователя__. Рекомендации на основе контента могут быть в значительной степени адаптированы к интересам пользователя, включая рекомендации по нишевым товарам, поскольку метод основан на сопоставлении характеристик или атрибутов объекта базы данных с интересами пользователя.
* __Рекомендации прозрачны для пользователя__. Высокорелевантные рекомендации создают ощущение понятности алгоритмов для пользователя, повышая уровень его доверия к предлагаемым рекомендациям.
* __Вы избегаете проблемы «холодного старта»__. Хотя фильтрация на основе контента требует первоначального ввода данных от пользователей, чтобы начать давать рекомендации, качество ранних рекомендаций обычно намного выше, чем у других подходов.
* __Системы фильтрации на основе содержания обычно проще в создании__. Основная работа заключается в создании характеристик, на основании которых будет вычисляться близость.

### Недостатки

* __Отсутствие новизны и разнообразия__. Вполне возможно, что человеку, который любит боевики, могли бы понравиться и фильмы ужасов. Однако мы не сможем ему их порекомендовать, если он сам не отметит подобные фильмы как понравившиеся.
* __Присвоенные характеристики могут быть неверными__. Рекомендации на основе контента хороши настолько, насколько хороши атрибуты (характеристики), присвоенные продуктам.

# Модель коллаборативной фильтрации

Эта модель основана на поиске __сходства между пользователями или между продуктами__. Используя этот метод, мы можем прогнозировать рейтинги на основе оценок похожих пользователей или похожих продуктов.

К примеру, если пользователь предпочитает фильмы ужасов, то ему можно посоветовать те фильмы, которые понравились другим поклонникам жанра «хоррор».

Поскольку рекомендации основаны на предпочтениях других пользователей, они называются коллаборативными. 

Существует два типа коллаборативной фильтрации: 
- на основе памяти (memory-based)
- на основе модели (model-based).

<img src='Images/math_ml_27.png'>

__Мерой близости__ может выступать косинусная близость между пользователями.

В целом, такой подход можно применять, однако у него есть ряд существенных __недостатков__:

* __Нечего рекомендовать новым/нетипичным пользователям__. Если появляется пользователь, который ни на кого не похож, мы не знаем, к какому кластеру его отнести.

* __Не учитывается специфика каждого пользователя__. По сути, мы выявляем некоторые паттерны поведения и предпочтений и для каждого паттерна выделяем свои рекомендации. Однако на самом деле даже пользователи из одного кластера немного отличаются друг от друга, поэтому возникают неточности.

* __Если оценок нет, то среднее арифметическое невозможно вычислить__. Если в кластере никто не оценивал объект, сделать предсказание не получится, так как для предсказания нужно вычислить среднее арифметическое для оценок.

# Коллаборативная фильтрация на основе памяти (memory-based)

Чтобы решить перечисленные выше проблемы, обратимся к коллаборативной фильтрации, а точнее к memory-based-подходу, основанному на близости пользователей (user-based).

Примечание: При __memory-based-подходе__ хранится полная матрица взаимодействий (лайков, просмотров и т.д.) пользователя с продуктом.

# Коллаборативная фильтрация на основе пользователей (User-based-подход)

Коллаборативная фильтрация на основе пользователей — это метод, используемый для предсказания продуктов, которые могут понравиться пользователю, на основе оценок, выставленных этому продукту другими пользователями, имеющими схожие с целевым пользователем вкусы.

__Коллаборативная фильтрация на основе пользователей__ — это метод, используемый для предсказания продуктов, которые могут понравиться пользователю, на основе оценок, выставленных этому продукту другими пользователями, имеющими схожие с целевым пользователем вкусы.

<img src='Images/math_ml_28.png'>

### Задача поиска схожести в user-based коллаборативной фильтрации

Цель: найти пользователей, которые оценивают продукты похоже, чтобы на основе их оценок рекомендовать товары.

Как решается:

1. У каждого пользователя есть вектор оценок товаров (строка в матрице предпочтений).

2. Считаем сходство между векторами пользователей (например, косинусное сходство, корреляция Пирсона).

3. Для данного пользователя выбираем наиболее схожих (top-N соседей).

4. Предлагаем товары, которые они высоко оценили, но которых нет у данного пользователя.

# Коллаборативная фильтрация на основе элементов (Item-based-подход)

Если мы транспонируем матрицу предпочтений и будем решать задачу сходства не для пользователей, а для объектов (items), то получим аналогичную задачу, которая является item-based-моделью коллаборативной фильтрации и даёт нам возможность предсказывать оценку.

Этот подход использует идею предыдущего, только теперь похожи не пользователи, а объекты. Возвращаясь к примеру с рекомендациями фильмов, теперь мы рекомендуем пользователю фильм, который похож на те фильмы, которые уже понравились этому пользователю ранее.

Кроме того, продуктов обычно больше, чем пользователей, поэтому векторы получатся большей размерности. Это даёт возможность получить более устойчивую модель с большей статистической значимостью.

### преимущества

* Может помочь пользователям обнаружить новые релевантные продукты из новых категорий, даже если пользователи не проявляют интерес к новым типам продуктов.
* Не требует подробных характеристик и контекстных данных о продуктах. По сути, для реализации нужна только матрица взаимодействий пользователей и продуктов.

### недостатки

* Нехватка данных может привести к трудностям при рекомендации новых продуктов или при появлении новых пользователей, поскольку предложения основаны на исторических данных и взаимодействии. Таким образом, для этого подхода актуальна проблема холодного старта.
* По мере роста пользовательской базы алгоритмы страдают из-за большого объёма данных (очень ресурсоёмкие вычисления) и недостаточной масштабируемости.
* Отсутствие разнообразия в долгосрочной перспективе. Это может показаться нелогичным, поскольку смысл коллаборативной фильтрации заключается в том, чтобы рекомендовать пользователю новые товары. Однако поскольку алгоритмы функционируют на основе исторических рейтингов, они не будут рекомендовать товары с небольшим количеством оценок или ограниченным количеством данных. Популярные товары будут более популярны в долгосрочной перспективе, а новых и разнообразных вариантов будет не хватать.

# Model-based-подход

В подходе на основе моделей используются модели машинного обучения для прогнозирования и ранжирования взаимодействий между пользователями и элементами, с которыми они ещё не взаимодействовали. Эти модели обучаются на основе информации, уже имеющейся в матрице взаимодействий, с помощью различных алгоритмов, например __матричной факторизации__.

__Матричная факторизация__ — это способ найти скрытые особенности пользователей и товаров, чтобы делать рекомендации.

### Что происходит?

* У нас есть большая таблица (матрица), где строки — это пользователи, а столбцы — товары.

* В ячейках — оценки товаров пользователями (обычно много пустых ячеек, потому что не все всё оценили).

### Что делает факторизация?

* Она разбивает эту большую таблицу на две маленькие таблицы:

-> Одна описывает каждого пользователя через несколько скрытых признаков (например, любовь к экшену, романтике и т.д.).

-> Другая описывает каждый товар через эти же скрытые признаки (например, насколько фильм — экшен, романтика и т.д.).

### Как это работает?

Если умножить эти две маленькие таблицы, мы получим приближение исходной большой таблицы (матрицы), где можно предсказать, как пользователь оценит новый товар.

### Пример:
Например, модель «узнала»:

* Пользователь 1 любит экшен (0.9) и не любит романтику (0.1).

-> Фильм X — это 0.8 экшен и 0.2 романтика.

-> Модель предскажет, что этот фильм должен понравиться этому пользователю.

## Рассмотрим два варианта матричной факторизации — SVD и ALS.

### SVD (Singular Value Decomposition)

Разлагает матрицу предпочтений на три матрицы:

M = U * S * $V^t$

* U — скрытые признаки пользователей

* S — диагональная матрица весов (сила скрытых признаков)

* $V^t$ — скрытые признаки товаров

Особенности:

* Обычный SVD требует полной матрицы (поэтому на практике используют модификации, типа TruncatedSVD).

Хорошо работает для уменьшения размерности и с плотными матрицами.

### ALS (Alternating Least Squares)

Итерирует:

* фиксируем признаки товаров → считаем признаки пользователей;

* фиксируем признаки пользователей → считаем признаки товаров.

Повторяем, пока ошибка не станет минимальной.

Особенности:

* Специально для разреженных матриц (много пропущенных оценок).

* Легко масштабируется на большие данные (работает с Spark).

Отлично подходит для коллаборативной фильтрации в продакшене.

### Итог:

1. __SVD__ — математически точный, но требует полной матрицы или её приближения.

2. __ALS__ — заточен под разреженные данные, удобен для больших масштабов.

# Гибридные модели

Выше мы рассмотрели отдельные модели для РС. Иногда эти модели используются вместе — такой способ построения рекомендательной системы называется __гибридным__.

__Гибридная РС__ — это особый тип рекомендательной системы, который представляет собой комбинацию из нескольких методов. Обычно это комбинация контентного подхода и коллаборативной фильтрации. Такое сочетание может помочь преодолеть недостатки, с которыми мы сталкиваемся при использовании этих методов по отдельности, а также в некоторых случаях может быть более эффективным.

<img src='Images/math_ml_29.png'>

Разумеется, можно комбинировать различные подходы самостоятельно, однако для удобства уже реализован __модуль LightFM__ — установить можно через следующую команду:

```python
    pip install lightfm
```

### преимущества 
* Использование разных моделей позволяет компенсировать их недостатки и использовать преимущества каждой модели для составления рекомендаций.
* Благодаря использованию сразу нескольких методов такой подход позволяет получить более персонализированные и точные рекомендации.

### недостатки
* Такие модели обычно имеют высокую вычислительную сложность и из-за этого долго обучаются.
* Для обучения таких моделей требуются большие объёмы данных и поступление новых данных для обновления рекомендаций.

# Глубокое обучение

__Глубокое обучение (Deep Learning, DL)__ — это современное и эффективное решение для многих задач машинного обучения, таких как компьютерное зрение или обработка естественного языка. Deep Learning во многих случаях превосходит классические методы, которые мы рассматривали ранее. Поэтому в последнее время глубокое обучение всё чаще применяется и в рекомендательных системах. Многие крупные компании, такие как AirBnB, Google, Home Depot, LinkedIn и Pinterest, используют рекомендательные системы, построенные именно на основе глубокого обучения.

### Преимущества использования нейронных сетей:

* Как правило, DL-модели дают более высокое качество. Стандартные ML-модели проигрывают глубокому обучению, особенно в ситуациях с большим объёмом данных.
* DL-модели обладают большей гибкостью. В рамках одной модели вы можете получить ответы на такие вопросы, как «Добавит ли пользователь товар в корзину?», «Начнёт ли он оформление заказа с этим товаром?» или «Купит ли он этот товар?».
* Можно включать в модель данные совершенно разных типов, в т.ч. текстовые данные (используя на них все инструменты NLP) или изображения (используя свёрточные нейронные сети).

### Концепция эмбеддинг

__Эмбеддинг__ — это пространство низкой размерности, которое отражает взаимосвязь векторов из пространства более высокой размерности.

Представим, что у нас есть различные данные о пользователях, которые мы хотим преобразовать в векторы с двумя координатами, отражающие всего две характеристики:

* степень симпатии к триллерам;
* степень симпатии к мелодрамам.

<img src='Images/math_ml_30.png'>

Теперь давайте представим некоторого пользователя Алису. Алиса предпочитает смотреть триллеры и не очень любит мелодрамы. Если в зависимости от её действий и оценок фильмов мы сможем оценить её любовь к триллерам и нелюбовь к мелодрамам в численном эквиваленте, то мы получим вектор, характеризующий предпочтения Алисы в плане жанров:

<img src='Images/math_ml_31.png'>

Рассмотрим ещё одного пользователя — Машу. Маша в целом является киноманом, так что примерно одинаково любит и триллеры, и мелодрамы. По аналогии с Алисой попробуем создать вектор, который характеризует предпочтения Маши.

<img src='Images/math_ml_32.png'>

Создав такие векторы-представления для всех пользователей, мы получим полный набор эмбеддингов.

По сути, эмбеддинги уменьшают размерность данных, оставляя осмысленность в их отображении. При таком преобразовании пользователи, предпочтения которых похожи, находятся рядом в плоскости (или пространстве), а пользователи, предпочтения которых отличаются, находятся далеко друг от друга.

Конечно, необязательно использовать для формирования эмбеддинга только два числа — для вектора можно взять любое количество компонент. Причём чем больше будет измерений, тем выше будет точность представления всех особенностей пользователя и, как следствие, выше будет точность модели. Однако следует понимать, что такое увеличение точности происходит за счёт __роста сложности__ модели и __увеличения количества времени__, необходимого для её обучения.

### Пример архитектуры нейронной сети

<img src='Images/math_ml_33.png'>

В качестве входных данных для нейронной сети мы передаём вектор характеристик пользователя и вектор для фильма (Inputs на схеме выше).

Из обоих этих векторов получаются эмбеддинги, о которых мы поговорили ранее (обозначены как Embedded User Vector и Embedded Item Vector). Далее эти эмбеддинги проходят через несколько полносвязных слоёв, на выходе из которых они преобразуются в вектор-предсказание. Например, на схеме выше видно, что вероятность первого класса (показывает, что элемент нерелевантен) — 0.2, а второго (показывает, что элемент релевантен) — 0.8. Следовательно, мы делаем выбор в пользу второго и рекомендуем этот продукт пользователю. Собственно, по такому алгоритму и обучается эта нейронная сеть.